### DISCUSSION TITLE
How to set/mock auth state in tests?

### DISCUSSION THREAD

**Discussion by litewarp** [gladly-team/next-firebase-auth#124] [General]
This could be my sleep-deprived brain misfiring, but how would one set the auth state on a rendering test for a component? 

The WithAuthUser HOC checks the return value of `useFirebaseUser`, and the tests for the package mock that function as well.

```
useFirebaseUser.mockReturnValue({
  user: createMockFirebaseUserClientSDK(),
  initialized: true,
})
```

However, because useFirebaseUser is not exported from the package, I can't seem to stub it properly in my app so my components wrapped in the HOC never render. Anyone got any ideas?

Thanks in advance.

**Comment by kmjennison** [comment]
We do this by mocking `next-firebase-auth` rather than worrying about its internal logic:

```js
// ./__mocks__/next-firebase-auth/withAuthUser.js
// the mock just renders its child component
const withAuthUser = jest.fn(() => (wrappedComponent) => wrappedComponent)
export default withAuthUser
```

```js
// ./__mocks__/next-firebase-auth/useAuthUser.js
import getMockAuthUser from './my/test/utils'
const useAuthUser = jest.fn(() => getMockAuthUser())
export default useAuthUser
```

Then, in tests, set the `useAuthUser` return value:
```js
useAuthUser.mockReturnValue(someValue)
```

If you wanted to, you could make your mocks more sophisticated so that the `withAuthUser` mock provides context to the `useAuthUser` mock, though you would probably end up replicating a lot of `next-firebase-auth` behavior.

Alternatively, if you chose not to mock `next-firebase-auth` at all, you'd want to mock the underlying Firebase libraries and likely also intercept network calls this package makes. This has the disadvantage of being much more complex and prone to breaking as this package changes, but it has the advantage of not implicitly trusting 3rd party code for critical auth logic.

**Reply by litewarp** [reply]
Thanks. That all makes sense. I'm still learning the ropes of testing React components, so this was helpful.

**Reply by litewarp** [reply]
Do you have a component test you could share? I've created the __mocks__ directory according to the [jest instructions](https://jestjs.io/docs/manual-mocks). 

But I keep getting a `TypeError: _nextFirebaseAuth.useAuthUser.mockReturnValue is not a function`

**Reply by kmjennison** [reply]
Sure, here's a repo where we use `next-firebase-auth`: [gladly-team/tab-web](https://github.com/gladly-team/tab-web)

Default mocks are [here](https://github.com/gladly-team/tab-web/tree/master/__mocks__/next-firebase-auth), and [here's a test](https://github.com/gladly-team/tab-web/blob/master/src/__tests__/pages/account.test.js#L136) where we use a mock return value for `useAuthUser`.

**Reply by grahaml** [reply]
Thanks very much for sharing the examples @kmjennison. FWIW it took me awhile to get this working, and it is kind of critical for any app using this library, as I expect I'll use these functions a lot. It might be worthwhile to add some documentation around this piece? I could try to have a crack at that. Would the right place to start be opening an issue for that?

**Reply by kmjennison** [reply]
@grahaml Sure, that sounds great! Yes, feel free to open an issue and take it on. Happy to add an "examples" section to the README.

**Reply by grahaml** [reply]
I'll see what I can do!

**Reply by grahaml** [reply]
Opened #133 to capture this. I'll try to tackle it over the next few days, while it's fresh in my mind.

**Reply by grahaml** [reply]
Had a spare cycle this afternoon so finessed something I wrote for my app's README to make it a bit more generic for this repo. See #134.
